src/wallet.rs

```diff
@@ -0,0 +1,942 @@
+                let mut prefix = [0u8; MAGIC.len() + VERSION.len()];
+                prefix[..MAGIC.len()].copy_from_slice(MAGIC);
+                prefix[MAGIC.len()..].copy_from_slice(VERSION);
+                content.splice(0..0, prefix.iter().cloned());
+
+                // write to file
+                fs::write(Path::new(&f.path().0), content)?;
+                Ok(())
+            }
+            None => Err(Error::WalletFileMissing),
+        }
+    }
+
+    /// Saves wallet to the provided file
+    pub fn save_to(&mut self, file: F) -> Result<(), Error> {
+        // set our new file
+        self.file = Some(file);
+        self.save()
+    }
+
+    /// Set a callback for status updates
+    pub fn set_status_callback(&mut self, func: fn(&str)) {
+        self.status = func
+    }
+
+    /// Connect the wallet to the network
+    pub async fn connect<R>(&mut self, endpoint: R) -> Result<(), Error>
+    where
+        R: RuskEndpoint,
+    {
+        // attempt connection
+        let rusk = RuskClient::connect(endpoint).await?;
+
+        // create a prover client
+        let prover = Prover::new(rusk.prover, rusk.state.clone(), rusk.network);
+
+        // create a state client
+        let state = State::new(rusk.state)?;
+
+        // create wallet instance
+        self.wallet = Some(WalletCore::new(self.store.clone(), state, prover));
+
+        Ok(())
+    }
+
+    /// Checks if the wallet has an active connection to the network
+    pub fn is_online(&self) -> bool {
+        self.wallet.is_some()
+    }
+
+    /// Obtain balance information for a given address
+    pub async fn get_balance(
+        &self,
+        addr: &Address,
+    ) -> Result<BalanceInfo, Error> {
+        // make sure we own this address
+        if !addr.is_owned() {
+            return Err(Error::Unauthorized);
+        }
+
+        // get balance
+        if let Some(wallet) = &self.wallet {
+            let index =
+                addr.index().expect("owned address should have an index");
+            wallet.get_balance(index).map_err(Error::from)
+        } else {
+            Err(Error::Offline)
+        }
+    }
+
+    /// Creates a new public address.
+    /// The addresses generated are deterministic across sessions.
+    pub fn new_address(&mut self) -> Address {
+        let addr = self.get_address(self.addrs.count);
+        self.addrs.count += 1;
+        addr
+    }
+
+    /// Default public address for this wallet
+    pub fn default_address(&self) -> Address {
+        let ssk = self
+            .store
+            .retrieve_ssk(0)
+            .expect("wallet seed should be available");
+        Address::new(0, ssk.public_spend_key())
+    }
+
+    fn get_address(&self, index: u64) -> Address {
+        let ssk = self
+            .store
+            .retrieve_ssk(index)
+            .expect("wallet seed should be available");
+        Address::new(self.address_count(), ssk.public_spend_key())
```

Comment:
Maybe this is a leftover from first refactor implementation
The address index should be the provided one instead of being the `address_count()`
```suggestion
        Address::new(index, ssk.public_spend_key())
```